lucet-runtime/lucet-runtime-internals/src/future.rs

```diff
@@ -0,0 +1,189 @@
+    /// make sure that it runs on a thread that is allowed to block.
+    ///
+    /// This argument is designed with [`tokio::task::block_in_place`][tokio] in mind. The odd type
+    /// is a concession to the fact that we don't have rank 2 types in Rust, and so must fall back
+    /// to trait objects in order to be able to take an argument that is itself a function that
+    /// takes a closure.
+    ///
+    /// In order to provide an appropriate function, you may have to wrap the library function in
+    /// another closure so that the types are compatible. For example:
+    ///
+    /// ```no_run
+    /// # async fn f() {
+    /// # let instance: lucet_runtime_internals::instance::InstanceHandle = unimplemented!();
+    /// fn block_in_place<F, R>(f: F) -> R
+    /// where
+    ///     F: FnOnce() -> R,
+    /// {
+    ///     // ...
+    ///     # f()
+    /// }
+    ///
+    /// instance.run_async("entrypoint", &[], |f| block_in_place(f)).await.unwrap();
+    /// # }
+    /// ```
+    ///
+    /// [tokio]: https://docs.rs/tokio/0.2.21/tokio/task/fn.block_in_place.html
+    pub async fn run_async<'a, F>(
+        &'a mut self,
+        entrypoint: &'a str,
+        args: &'a [Val],
+        wrap_blocking: F,
+    ) -> Result<RunResult, Error>
+    where
+        F: Fn(&mut (dyn FnMut() -> Result<RunResult, Error>)) -> Result<RunResult, Error>,
+    {
+        if self.is_yielded() {
+            return Err(Error::Unsupported(
+                "cannot run_async a yielded instance".to_owned(),
+            ));
+        }
+
+        // Store the ResumeVal here when we get it.
+        let mut resume_val: Option<ResumeVal> = None;
+        let ret = loop {
+            // Run the WebAssembly program
+            let run = wrap_blocking(&mut || {
+                if self.is_yielded() {
+                    // A previous iteration of the loop stored the ResumeVal in
+                    // `resume_val`, send it back to the guest ctx and continue
+                    // running:
+                    self.resume_with_val_impl(
+                        resume_val
+                            .take()
+                            .expect("is_yielded implies resume_value is some"),
+                        true,
+                    )
+                } else {
+                    // This is the first iteration, call the entrypoint:
+                    let func = self.module.get_export_func(entrypoint)?;
+                    self.run_func(func, args, true)
+                }
+            });
+            match run {
+                Ok(run_result) => {
+                    // Check if the yield came from Vmctx::block_on:
+                    if run_result.has_yielded::<YieldedFuture>() {
+                        let YieldedFuture(future) = *run_result
+                            .yielded()
+                            .unwrap()
+                            .downcast::<YieldedFuture>()
+                            .unwrap();
```

Comment:
Instead of these unwraps, how about adding a method:

```rust
impl RunResult {
    fn downcast_yielded<T: Any + 'static>(self) -> Result<T, Self> { ... }
}
```